ORStools/proc/export_proc.py

```diff
@@ -0,0 +1,168 @@
+ ***************************************************************************/
+"""
+
+from typing import Dict
+
+from qgis.core import (
+    QgsWkbTypes,
+    QgsFeature,
+    QgsField,
+    QgsFields,
+    QgsProject,
+    QgsCoordinateTransform,
+    QgsCoordinateReferenceSystem,
+    QgsProcessingParameterExtent,
+    QgsProcessingContext,
+    QgsProcessingFeedback,
+    QgsPointXY,
+    QgsGeometry,
+)
+
+from qgis.PyQt.QtCore import QVariant
+
+from qgis.utils import iface
+
+
+from ORStools.common import PROFILES
+from ORStools.utils import exceptions, logger
+from .base_processing_algorithm import ORSBaseProcessingAlgorithm
+
+
+# noinspection PyPep8Naming
+class ORSExportAlgo(ORSBaseProcessingAlgorithm):
+    def __init__(self):
+        super().__init__()
+        self.ALGO_NAME: str = "export_network_from_map"
+        self.GROUP: str = "Export"
+        self.IN_EXPORT: str = "INPUT_EXPORT"
+        self.PARAMETERS: list = [
+            QgsProcessingParameterExtent(
+                name=self.IN_EXPORT,
+                description=self.tr("Input Extent"),
+            ),
+        ]
+
+    def processAlgorithm(
+        self, parameters: dict, context: QgsProcessingContext, feedback: QgsProcessingFeedback
+    ) -> Dict[str, str]:
+        ors_client = self._get_ors_client_from_provider(parameters[self.IN_PROVIDER], feedback)
+
+        # Get profile value
+        profile = dict(enumerate(PROFILES))[parameters[self.IN_PROFILE]]
+
+        rect = self.parameterAsExtent(parameters, self.IN_EXPORT, context)
+
+        target_crs = QgsCoordinateReferenceSystem("EPSG:4326")
+        source_crs = iface.mapCanvas().mapSettings().destinationCrs()
+
+        transform = QgsCoordinateTransform(source_crs, target_crs, QgsProject.instance())
+
+        bottom_left = transform.transform(rect.xMinimum(), rect.yMinimum())
+        top_right = transform.transform(rect.xMaximum(), rect.yMaximum())
+
+        extent = [[bottom_left.x(), bottom_left.y()], [top_right.x(), top_right.y()]]
+
+        params = {
+            "bbox": extent,
+            "id": "export_request",
+        }
+
+        sink_fields = self.get_fields()
+
+        (sink, dest_id) = self.parameterAsSink(
+            parameters,
+            self.OUT,
+            context,
+            sink_fields,
+            QgsWkbTypes.Type.LineString,
+            QgsCoordinateReferenceSystem.fromEpsgId(4326),
+        )
+
+        # Make request and catch ApiError
+        try:
+            response = ors_client.request("/v2/export/" + profile, {}, post_json=params)
+            edges = response["edges"]
+            for edge in edges:
+                from_id = edge["fromId"]
+                to_id = edge["toId"]
+                weight = edge["weight"]
+
+                to_coords = self.get_location_by_id(to_id, response["nodes"])
+                from_coords = self.get_location_by_id(from_id, response["nodes"])
```

Comment:
This iterates all nodes twice for every edge, taking what is in theory a linear operation and making it quadratic.

I think it makes sense to parse nodes into a dict to look up coordinates easily.